File: acuity/libraries/templates/AcuityClient.java
// AcuityClient.java
//
// Drop-in helper for reading from an Acuity vision coprocessor over
// NetworkTables 4. Subscribes to the /acuity/* topics the device's
// dashboard publishes; gives robot code a typed Optional<TagDetection>
// so they don't have to deal with raw NT4 calls.
//
// Zero external dependencies — just WPILib's bundled NT4. No vendordep
// to install, no Maven artifact to fetch. Manager dropped this file
// in for you; if you need to update it, re-run the Libraries tab.
//
// Usage:
//
//     import frc.robot.acuity.AcuityClient;
//
//     private final AcuityClient acuity = AcuityClient.getInstance();
//
//     @Override
//     public void teleopPeriodic() {
//       acuity.getBestTag().ifPresent(tag -> {
//         shooter.aim(tag.yawDeg(), tag.distanceMeters());
//       });
//     }
//
// Schema reference: see your Acuity device's Docs tab → "NetworkTables
// schema", or acuity/docs/nt4-schema.md in the firmware repo.

package frc.robot.acuity;

import edu.wpi.first.networktables.DoubleSubscriber;
import edu.wpi.first.networktables.IntegerSubscriber;
import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.NetworkTableInstance;

import java.util.Optional;

public final class AcuityClient {

  /** Single tag detection from the device's "best target" topic. */
  public static record TagDetection(
      int id,
      double distanceMeters,
      double yawDeg,
      double pitchDeg,
      double tx,
      double ty,
      double area,
      double timestamp,
      double decisionMargin) {}

  /** On-device system health, mirrored from /acuity/health/*. */
  public static record Health(double cpuPct, double tempC, long uptimeS) {}

  // Heartbeat-staleness window. NT4 itself is push-based, so the
  // useful question isn't "is the wire connected" but "have we
  // received a fresh value lately?". 250 ms is comfortably bigger
  // than one frame at the slowest expected publish rate (10 Hz).
  private static final long STALE_MS = 250;

  private static AcuityClient instance;

  /** Process-wide singleton. Call this from robotInit / RobotPeriodic. */
  public static AcuityClient getInstance() {
    if (instance == null) instance = new AcuityClient();
    return instance;
  }

  private final IntegerSubscriber heartbeat;
  private final IntegerSubscriber bestId;
  private final DoubleSubscriber  bestDistanceM;
  private final DoubleSubscriber  bestYawDeg;
  private final DoubleSubscriber  bestPitchDeg;
  private final DoubleSubscriber  bestTx;
  private final DoubleSubscriber  bestTy;
  private final DoubleSubscriber  bestArea;
  private final DoubleSubscriber  bestTimestamp;
  private final DoubleSubscriber  bestDecisionMargin;
  private final DoubleSubscriber  cpuPct;
  private final DoubleSubscriber  tempC;
  private final IntegerSubscriber uptimeS;

  // We tell "stale" from "fresh" by watching the heartbeat int change.
  // The dashboard increments it once per published frame; if we keep
  // reading the same value, the device has gone away.
  private long lastHeartbeatValue = Long.MIN_VALUE;
  private long lastHeartbeatChangeMs = 0;

  private AcuityClient() {
    NetworkTable t = NetworkTableInstance.getDefault().getTable("acuity");
    heartbeat          = t.getIntegerTopic("heartbeat").subscribe(0);
    bestId             = t.getIntegerTopic("tags/best/id").subscribe(-1);
    bestDistanceM      = t.getDoubleTopic("tags/best/distance_m").subscribe(0.0);
    bestYawDeg         = t.getDoubleTopic("tags/best/yaw_deg").subscribe(0.0);
    bestPitchDeg       = t.getDoubleTopic("tags/best/pitch_deg").subscribe(0.0);
    bestTx             = t.getDoubleTopic("tags/best/tx").subscribe(0.0);
    bestTy             = t.getDoubleTopic("tags/best/ty").subscribe(0.0);
    bestArea           = t.getDoubleTopic("tags/best/area").subscribe(0.0);
    bestTimestamp      = t.getDoubleTopic("tags/best/timestamp").subscribe(0.0);
    bestDecisionMargin = t.getDoubleTopic("tags/best/decision_margin").subscribe(0.0);
    cpuPct             = t.getDoubleTopic("health/cpu_pct").subscribe(0.0);
    tempC              = t.getDoubleTopic("health/temp_c").subscribe(0.0);
    uptimeS            = t.getIntegerTopic("health/uptime_s").subscribe(0);
  }

  /** True when the device has published an updated heartbeat in the
   *  last 250 ms. Use this to gate display logic ("device offline"). */
  public boolean isConnected() {
    long h = heartbeat.get();
    long now = System.currentTimeMillis();
    if (h != lastHeartbeatValue) {
      lastHeartbeatValue = h;
      lastHeartbeatChangeMs = now;
    }
    return lastHeartbeatChangeMs > 0
        && (now - lastHeartbeatChangeMs) < STALE_MS;
  }

  /** Best target in current frame, or empty if none / device offline.
   *  "Best" is whichever tag the dashboard's selection logic picked
   *  (operator-locked tag → preferred-list match → largest tag). */
  public Optional<TagDetection> getBestTag() {
    if (!isConnected()) return Optional.empty();
    int id = (int) bestId.get();
    if (id < 0) return Optional.empty();
    return Optional.of(new TagDetection(
        id,
        bestDistanceM.get(),
        bestYawDeg.get(),
        bestPitchDeg.get(),
        bestTx.get(),
        bestTy.get(),
        bestArea.get(),
        bestTimestamp.get(),
        bestDecisionMargin.get()));
  }

  /** SoC stats — mostly useful for telemetry overlays. */
  public Health getHealth() {
    return new Health(cpuPct.get(), tempC.get(), uptimeS.get());
  }
}

File: acuity/libraries/templates/acuity_client.py
"""acuity_client.py — drop-in helper for talking to an Acuity vision
coprocessor over NetworkTables 4 from robotpy.

Subscribes to /acuity/* topics published by the on-device dashboard;
gives robot code a typed `TagDetection` so they don't have to deal
with raw NT4 calls.

Zero pip dependencies beyond robotpy's bundled `ntcore`. Manager
dropped this file in for you; if you want to update it, re-run the
Libraries tab.

Usage:

    from acuity_client import AcuityClient

    class MyRobot(wpilib.TimedRobot):
        def robotInit(self):
            self.acuity = AcuityClient()

        def teleopPeriodic(self):
            tag = self.acuity.get_best_tag()
            if tag is not None:
                self.shooter.aim(tag.yaw_deg, tag.distance_m)

Schema reference: see your Acuity device's Docs tab → "NetworkTables
schema", or acuity/docs/nt4-schema.md in the firmware repo.
"""
from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import List, Optional

import ntcore


@dataclass(frozen=True)
class TagDetection:
    id: int
    distance_m: float
    yaw_deg: float
    pitch_deg: float
    tx: float
    ty: float
    area: float
    timestamp: float
    decision_margin: float


@dataclass(frozen=True)
class Health:
    cpu_pct: float
    temp_c: float
    uptime_s: int


class AcuityClient:
    """Single-instance NT4 reader for an Acuity device.

    Construct one of these in robotInit and reuse it. Cheap to call
    `get_best_tag()` every periodic tick — it's just NT4 reads.
    """

    # Heartbeat-staleness window (ms). NT4 is push-based, so the
    # useful "is the device alive" check is "did we receive a fresh
    # heartbeat lately?" rather than a wire-level connection probe.
    STALE_MS = 250

    def __init__(self):
        inst = ntcore.NetworkTableInstance.getDefault()
        t = inst.getTable("acuity")
        self._heartbeat = t.getIntegerTopic("heartbeat").subscribe(0)
        self._best_id   = t.getIntegerTopic("tags/best/id").subscribe(-1)
        self._best_dist = t.getDoubleTopic("tags/best/distance_m").subscribe(0.0)
        self._best_yaw  = t.getDoubleTopic("tags/best/yaw_deg").subscribe(0.0)
        self._best_pitch = t.getDoubleTopic("tags/best/pitch_deg").subscribe(0.0)
        self._best_tx   = t.getDoubleTopic("tags/best/tx").subscribe(0.0)
        self._best_ty   = t.getDoubleTopic("tags/best/ty").subscribe(0.0)
        self._best_area = t.getDoubleTopic("tags/best/area").subscribe(0.0)
        self._best_ts   = t.getDoubleTopic("tags/best/timestamp").subscribe(0.0)
        self._best_dm   = t.getDoubleTopic("tags/best/decision_margin").subscribe(0.0)
        self._all_tags  = t.getStringTopic("tags/all").subscribe("[]")
        self._cpu_pct   = t.getDoubleTopic("health/cpu_pct").subscribe(0.0)
        self._temp_c    = t.getDoubleTopic("health/temp_c").subscribe(0.0)
        self._uptime    = t.getIntegerTopic("health/uptime_s").subscribe(0)

        # Heartbeat-change tracker. `_last_hb_change_ms == 0` means
        # "we've never seen a fresh value yet" — distinct from
        # "last fresh value was a long time ago".
        self._last_hb_value = -1
        self._last_hb_change_ms = 0

    def is_connected(self) -> bool:
        """True if the device has published a fresh heartbeat in the
        last STALE_MS milliseconds."""
        h = int(self._heartbeat.get())
        now_ms = int(time.monotonic() * 1000)
        if h != self._last_hb_value:
            self._last_hb_value = h
            self._last_hb_change_ms = now_ms
        return (self._last_hb_change_ms > 0
                and (now_ms - self._last_hb_change_ms) < self.STALE_MS)

    def get_best_tag(self) -> Optional[TagDetection]:
        """Best target in current frame, or None if none / device
        offline."""
        if not self.is_connected():
            return None
        id_ = int(self._best_id.get())
        if id_ < 0:
            return None
        return TagDetection(
            id=id_,
            distance_m=float(self._best_dist.get()),
            yaw_deg=float(self._best_yaw.get()),
            pitch_deg=float(self._best_pitch.get()),
            tx=float(self._best_tx.get()),
            ty=float(self._best_ty.get()),
            area=float(self._best_area.get()),
            timestamp=float(self._best_ts.get()),
            decision_margin=float(self._best_dm.get()),
        )

    def get_all_tags(self) -> List[TagDetection]:
        """Every tag visible in the latest frame. The all-tags topic
        is JSON-encoded — we parse it here so callers don't have to."""
        if not self.is_connected():
            return []
        try:
            data = json.loads(self._all_tags.get() or "[]")
        except (ValueError, TypeError):
            return []
        out: List[TagDetection] = []
        for d in data:
            if not isinstance(d, dict):
                continue
            try:
                out.append(TagDetection(
                    id=int(d.get("id", -1)),
                    distance_m=float(d.get("distance_m", 0.0)),
                    yaw_deg=float(d.get("yaw_deg", 0.0)),
                    pitch_deg=float(d.get("pitch_deg", 0.0)),
                    tx=float(d.get("tx", 0.0)),
                    ty=float(d.get("ty", 0.0)),
                    area=float(d.get("area", 0.0)),
                    timestamp=float(d.get("timestamp", 0.0)),
                    decision_margin=float(d.get("decision_margin", 0.0)),
                ))
            except (TypeError, ValueError):
                continue
        return out

    def get_health(self) -> Health:
        return Health(
            cpu_pct=float(self._cpu_pct.get()),
            temp_c=float(self._temp_c.get()),
            uptime_s=int(self._uptime.get()),
        )
